---
title: "Stickleback TagSeq Transcriptomics Analyses Walkthrough"
output: html_document
---
  
  This is an [R Markdown](http://rmarkdown.rstudio.com) walkthrough of the data analyses associated with **PAPER**. Follow the step by step instructions below to go from supplied data to results presented in **PAPER**. Source codes and raw data are avilable via [GitHub](https://github.com/lfuess/TagSeqMS).

##Preparing the Files

We start by preparing files for input into DESeq2 and subsequent analyses. We need to first parse out our metadata file from the original file that was generated during experimentation. Then we need to parse out our read count matrix generated from our TagSeq analysis pipeline to remove any samples that are missing values in our metadata. The following code will parse out both files and output the correct input for subsequent analyses: **Generate_InputFiles.R**

```{r echo = FALSE, warning = FALSE, message = FALSE }
source("Generate_InputFiles.R")
```

##Differential Expression Analysis

The next step is to run the differential expression analyses. This takes quite some time to run but will output the necessary files for futher pathway analysis. Here is the code for DESeq2: **Differential_Expression.R**

Once we've run DESeq2, we can parse the resulting files to get lists of significant genes for each model factor, and to create an input file for Ingenuity Pathway Analysis. The following file will do just that: **ParseDESeq2Results.R**

```{r echo = FALSE, warning = FALSE, message = FALSE }
source("ParseDESeq2Results.R")
```

##Pathway Analysis

The next steps are all done in Ingenuity Pathway Analysis and via hand curration. The associated output files can be found in the GitHub and Supplementary Materials. We'll use these output files to make some of the following figures.

##Figure 1

This figure depicts the output of Ingenuity Pathway Analysis for the effects of infection on pathway activation. It uses a hand currated file resulting from the output of IPA. The code is found in: **Figure1.R**

```{r fig1, echo=FALSE}
library(ggplot2)

infect = read.csv("Infect_bar_all.csv")
ggplot(infect, aes(y=z.score, x=reorder(Pathway, z.score))) + 
  geom_bar(position="dodge", stat="identity", fill = "#cc3d24") + coord_flip()+
  theme_classic() + geom_abline(slope=0, intercept=0,  col = "black")+
  xlab("Pathway") + ylab("z-score")
```

##Figure 2

This figure requires a little more detail; First we need to run a code that will help us look at overlap in genes affected by our different model factors. This code is found here: **Compare_Effects.R**

We can use the resulting files to create our Venn Diagram input. Final code is found at: **Figure2.R**

```{r fig2, echo=FALSE, message=FALSE, warning=FALSE}
library(VennDiagram)

draw.triple.venn(area1=7745, area2=10601, area3=1202, n12=7175, n13=399, n23=856, n123=203, category = c("F2vGBC", "F2vRBC", "RBCvGBC"), lty = "blank", fill=c("#6dae90", "#f3c558", "#30b4cc"), alpha=rep(.7, 3), cex=rep(1.7, 7), fontfamily=rep("Arial Black", 7), cat.cex=rep(1, 3), cat.fontface=rep("bold.italic", 3), cat.fontfamily=rep("Arial",3))

```

##Figure 3

Figure 3 shows patterns of expression for genes that respond in different ways to infection in RBC vs. GBC fish. We use a single code to parse the data and generate this figure. Final code is found at: **Figure3.R**

```{r fig3, echo=FALSE, message=FALSE, warning=FALSE}
##lets make figure 3 for the MS##
library(tidyverse)
library(data.table)

##first we have to process the data##
  ##input reads##
  reads = read.csv("normalizedreads.csv", check.names = FALSE)
  colnames(reads)[1] <- "Gene"
  row.names(reads) <- reads$Gene
  reads[1] <- NULL
  ##select sig genes##
  sig = reads[c("ENSGACT00000015170.1", "ENSGACT00000018024.1","ENSGACT00000024555.1","ENSGACT00000003040.1"),]
  sig = as.data.frame(t(sig))
  sig=setDT(sig, keep.rownames = TRUE)[]
  colnames(sig)[1] <- "Sample"
  ##input metadata##
  meta = read.csv("ExpDesign.csv")
  meta = meta[,c(1,3,6)]
  ##select RBC##
  RBC= meta[meta$CrossDir == "RBC", ]
  RBC_Infect = RBC[RBC$worm_present == "TRUE", ]  
  RBC_Infect <- as.vector(RBC_Infect['Sample'])
  RBC_UnInfect = RBC[RBC$worm_present == "FALSE", ]  
  RBC_UnInfect <- as.vector(RBC_UnInfect['Sample'])
  ##Select GBC##
  GBC= meta[meta$CrossDir == "GBC", ]
  GBC_Infect = GBC[GBC$worm_present == "TRUE", ]  
  GBC_Infect <- as.vector(GBC_Infect['Sample'])
  GBC_UnInfect = GBC[GBC$worm_present == "FALSE", ]  
  GBC_UnInfect <- as.vector(GBC_UnInfect['Sample'])
  
##now we can subset the data to create averages and stderr##
  RBC_Infect = merge(sig,RBC_Infect, by="Sample")
  RBC_Infect = RBC_Infect %>% remove_rownames %>% column_to_rownames(var="Sample")
  RBC_Infect[c("average"),]=colMeans(RBC_Infect, na.rm = FALSE, dims = 1)  
  RBC_Infect[c("se"),]=sapply(RBC_Infect,function(x)sd(x)/sqrt(length(x)))
  RBC_Infect[c("Cross"),]="RBC"  
  RBC_Infect[c("Infect"),]="TRUE"
  RBC_Infect[nrow(RBC_Infect) + 1,] = c("CCN3","Unannotated","SH2D1A","LSM10")
  rownames(RBC_Infect)[29]<-"name"
  RBC_Infect=RBC_Infect[c(25:29),]  
  RBC_Infect=as.data.frame(t(RBC_Infect))
  
  
  GBC_Infect = merge(sig,GBC_Infect, by="Sample")
  GBC_Infect = GBC_Infect %>% remove_rownames %>% column_to_rownames(var="Sample")
  GBC_Infect[c("average"),]=colMeans(GBC_Infect, na.rm = FALSE, dims = 1)  
  GBC_Infect[c("se"),]=sapply(GBC_Infect,function(x)sd(x)/sqrt(length(x)))
  GBC_Infect[c("Cross"),]="GBC"  
  GBC_Infect[c("Infect"),]="TRUE"
  GBC_Infect[nrow(GBC_Infect) + 1,] = c("CCN3","Unannotated","SH2D1A","LSM10")
  rownames(GBC_Infect)[25]<-"name"
  GBC_Infect=GBC_Infect[c(21:25),]  
  GBC_Infect=as.data.frame(t(GBC_Infect))
  
  
  
  RBC_UnInfect = merge(sig,RBC_UnInfect, by="Sample")
  RBC_UnInfect = RBC_UnInfect %>% remove_rownames %>% column_to_rownames(var="Sample")
  RBC_UnInfect[c("average"),]=colMeans(RBC_UnInfect, na.rm = FALSE, dims = 1)  
  RBC_UnInfect[c("se"),]=sapply(RBC_UnInfect,function(x)sd(x)/sqrt(length(x)))
  RBC_UnInfect[c("Cross"),]="RBC"  
  RBC_UnInfect[c("Infect"),]="FALSE"
  RBC_UnInfect[nrow(RBC_UnInfect) + 1,] = c("CCN3","Unannotated","SH2D1A","LSM10")
  rownames(RBC_UnInfect)[73]<-"name"
  RBC_UnInfect=RBC_UnInfect[c(69:73),]  
  RBC_UnInfect=as.data.frame(t(RBC_UnInfect))
  
  
  GBC_UnInfect = merge(sig,GBC_UnInfect, by="Sample")
  GBC_UnInfect = GBC_UnInfect %>% remove_rownames %>% column_to_rownames(var="Sample")
  GBC_UnInfect[c("average"),]=colMeans(GBC_UnInfect, na.rm = FALSE, dims = 1)  
  GBC_UnInfect[c("se"),]=sapply(GBC_UnInfect,function(x)sd(x)/sqrt(length(x)))
  GBC_UnInfect[c("Cross"),]="GBC"  
  GBC_UnInfect[c("Infect"),]="FALSE"
  GBC_UnInfect[nrow(GBC_UnInfect) + 1,] = c("CCN3","Unannotated","SH2D1A","LSM10")
  rownames(GBC_UnInfect)[70]<-"name"
  GBC_UnInfect=GBC_UnInfect[c(66:70),]  
  GBC_UnInfect=as.data.frame(t(GBC_UnInfect))

  ##merge them all together to get your final data frame for graphing!!
  data = rbind(RBC_Infect, RBC_UnInfect, GBC_Infect, GBC_UnInfect)

  
##Actual Graphing Part##
  levels(data$Infect)[levels(data$Infect)=="FALSE"] <- "Uninfected"
  levels(data$Infect)[levels(data$Infect)=="TRUE"] <- "Infected"
  data$se=as.numeric(as.character(data$se))
  data$average=as.numeric(as.character(data$average))

  data$Infect <- factor(data$Infect, levels = rev(levels(factor(data$Infect))))
  
  gp1 <- ggplot(data, aes(x=Infect, y=average, colour=Cross, group=Cross))
  gp1 + geom_line(aes(linetype=Cross), size=.6) + scale_colour_manual(values=c("#f3c558","#6dae90")) +
    geom_point(aes(shape=Cross), size=3) + scale_linetype_manual(values=c("solid", "solid")) + 
    scale_shape_manual(values=c(17, 18)) + geom_errorbar(aes(ymax=average+se, ymin=average-se), width=.1) + 
    facet_wrap(~name, ncol=2, scales="free") + theme_bw() +
    theme(panel.border = element_blank(), panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
    ylab("Normalized Expression") + xlab("Infection") 

```

##Figure 4

This is very similar to Figure 1, but it looks specifically at Fibrosis rather than infection. Code is found at: **Figure4.R**

```{r fig4, echo=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)

fibrosis = read.csv("Fib_bar_all.csv")
ggplot(fibrosis, aes(y=z.score, x=reorder(Pathway, z.score))) + 
  geom_bar(position="dodge", stat="identity", fill = "#004f7a" ) + coord_flip()+
  theme_classic() + geom_abline(slope=0, intercept=0,  col = "black")+
  xlab("Pathway") + ylab("z-score")

```

##Figure 5

The final figure compares pathways which were shared across infection and fibrosis, looking specifically at their comparative activation scores (z-scores). Using a scatter plot we can show distribution of pathways to evaluate which are more or less activated as a result of infection vs fibrosis. It also uses a hand-currated file based on output from IPA. Code is at: **Figure5.R**

```{r fig5, echo=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)

scat = read.csv("Infect_Fib_path_scat.csv")

ggplot(scat, aes(y=Fibrosis, x=Infection, color=Color)) + 
  geom_point() + geom_abline(slope=1, intercept=0, linetype="dashed") +
  theme_classic() +
  theme(legend.background = element_rect(fill="white",
                                         size=0.5, linetype="solid", 
                                         colour ="black")) +
  scale_color_manual(values=c("#004f7a", "#cc3d24")) +
  xlab("Infection z-score") + ylab("Fibrosis z-score") +
  theme(legend.position="none")
```

##Supplementary Figure 1

Supplementary Figure 1 is a combination of Figures 1, 4, and 5, but instead of pathways it displays results for upstream regulator analyses. Code is at: **SupFig1.R**

```{r supfig1, echo=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)

##let's make a regulator comparison bar graph

library(ggplot2)

scat = read.csv("Infect_Fib_reg_scat.csv")

p1 = ggplot(scat, aes(y=Fibrosis, x=Infection, color=Color)) + 
  geom_point() + geom_abline(slope=1, intercept=0, linetype="dashed") +
  theme_classic() +
  theme(legend.background = element_rect(fill="white",
                                         size=0.5, linetype="solid", 
                                         colour ="black")) +
  scale_color_manual(values=c("#004f7a", "#cc3d24")) +
  xlab("Infection z-score") + ylab("Fibrosis z-score") +
  theme(legend.position="none")

##we can also make unique plots for each fibrosis and infection##

infect = read.csv("Infect_unique_reg_bar.csv")
p2 = ggplot(infect, aes(y=z.score, x=reorder(Regulator, z.score))) + 
  geom_bar(position="dodge", stat="identity", fill = "#cc3d24") + coord_flip()+
  theme_classic() + geom_abline(slope=0, intercept=0,  col = "black")+
  xlab("Regulator") + ylab("z-score")


fibrosis = read.csv("Fib_unique_reg_bar.csv")
p3 = ggplot(fibrosis, aes(y=z.score, x=reorder(Regulator, z.score))) + 
  geom_bar(position="dodge", stat="identity", fill = "#004f7a" ) + coord_flip()+
  theme_classic() + geom_abline(slope=0, intercept=0,  col = "black")+
  xlab("Regulator") + ylab("z-score")

##make them into a multiplot##
library("cowplot")
p4 = plot_grid(p2, p3, ncol = 1, align = "v")
ggdraw() +
  draw_plot(p4, x = 0, y = 0, width = .5, height = 1) +
  draw_plot(p1, x = .5, y = .5, width = .5, height = .5) +
  draw_plot_label(label = c("A", "C", "B"), size = 15,
                  x = c(0, 0.5, 0), y = c(1, 1, 0.5))
```

##Supplementary Figure 2

Finally, this figure is similar to Figure 2. Based on a hand curated file from the IPA output, we construct a Venn Diagram of overlap in pathways which are differentially activated as a result of each of our three cross-type comparisons. Code is found at: **SuppFig2.R**

```{r supfig2, echo=FALSE, message=FALSE, warning=FALSE}
library(VennDiagram)

draw.triple.venn(area1=240, area2=313, area3=94, n12=231, n13=74, n23=85, n123=74, category = c("F2vGBC", "F2vRBC", "RBCvGBC"), lty = "blank", fill=c("#6dae90", "#f3c558", "#30b4cc"), alpha=rep(.7, 3), cex=rep(1.7, 7), fontfamily=rep("Arial Black", 7), cat.cex=rep(1, 3), cat.fontface=rep("bold.italic", 3), cat.fontfamily=rep("Arial",3))

```